### PR TITLE
[RFC] Support for registryAlias in Association

### PR DESCRIPTION
**Pull request by robertpustulka**
See #10140

### CODE REVIEW

**Comment by robertpustulka** [comment]
I'm wondering if a `registryAlias` setter should be allowed. Changing it after the target table has been loaded would have no effect. `className` option is immutable.

File: src/ORM/Association.php [hunk 1]
```diff
@@ -366,6 +374,25 @@ public function source(Table $table = null)
...
 
     /**
+     * The registry alias of the target table object
+     *
+     * @return string
+     */
+    public function getRegistryAlias()
+    {
+        if ($this->_registryAlias === null) {
+            if (strpos($this->_className, '.')) {
```
Comment (markstory) [inline src/ORM/Association.php]: Do you care if the `.` is in the first char?

File: tests/TestCase/ORM/EagerLoaderTest.php [hunk 2]
```diff
@@ -413,6 +413,28 @@ public function testContainToFieldsDefault()
...
     /**
+     * Tests that custom registry alias is honored in contain.
+     *
+     * @return void
+     */
+    public function testContainWithCustomRegistryAlias()
+    {
+        $this->table->belongsTo('foo', [
+            'registryAlias' => 'clients'
+        ]);
```
Comment (markstory) [inline tests/TestCase/ORM/EagerLoaderTest.php]: It would be good to have tests for matching as well.

File: src/ORM/Association.php (hunk 1 above)
Comment (robertpustulka) [inline src/ORM/Association.php]: @markstory It's just a copy-paste from an old `getTargetTable()` method.

**Comment by robertpustulka** [comment]
I'm working on ensuring that container results are transformed to entites. My approach is to use `registryAlias` as a table alias in queries.

Found out that for this to work I had to replace occurences of association name in joins, `aliasPath` and `nestKey` options with target table alias.

It went well to the point where BTM associations stopped working. I found out that it's because of join data (see failed tests).

There's an exception: `RuntimeException: "_joinData" is missing from the belongsToMany results. Results cannot be created.` [Here is the code responsible.](https://github.com/robertpustulka/cakephp/blob/associations-registry-alias/src/ORM/Association/Loader/SelectWithPivotLoader.php#L171). 

And this is how the sample failing data looks:

```php
object(Cake\ORM\Entity) {

	'id' => (int) 1,
	'name' => 'tag1',
	'description' => 'A big description',
	'created' => object(Cake\I18n\Time) {

		'time' => '2016-01-01T00:00:00+00:00',
		'timezone' => 'UTC',
		'fixedNowTime' => '2017-02-01T12:59:32+00:00'
	
	},
	'Tags_CJoin' => [
		'article_id' => (int) 1,
		'tag_id' => (int) 1
	],
	'[new]' => false,
	'[accessible]' => [
		'*' => true
	],
	'[dirty]' => [],
	'[original]' => [],
	'[virtual]' => [],
	'[errors]' => [],
	'[invalid]' => [],
	'[repository]' => 'Tags'

}
```

Seems like temporary association data is not replaced with `_joinData`. Probably because of messing with join names.

Could you help me with that? I'm a bit stuck here.

I'm not sure if I should pursue this or change my approach (use association name in queries) or drop the idea...

**Comment by lorenzo** [comment]
Why don't we use the registry alias as the table alias itself?

**Comment by lorenzo** [comment]
> Why don't we use the registry alias as the table alias itself?

Just answered myself this question, this would require you to write conditions and fields by prefixing with the registry alias as well. 

I think this idea is immediately revealing the original constraints that forced the current design.

File: src/ORM/Association.php [hunk 3]
```diff
@@ -366,6 +374,25 @@ public function source(Table $table = null)
     }
 
     /**
+     * The registry alias of the target table object
+     *
+     * @return string
+     */
+    public function getRegistryAlias()
```
Comment (markstory) [inline src/ORM/Association.php]: Should 'target' be in this method's name? It is the alias of the target table after all.

**Comment by robertpustulka** [comment]
@lorenzo so what are you suggesting?

**Comment by lorenzo** [comment]
@robertpustulka maybe back to the drawing board. What is is the problem we are trying to solve and what are the least of all constraints that make achieving that goal more difficult?

As I understand you want to have repeated aliases names for associations. These are the problems we need to solve:

* What happens if someone adds the same association name twice in the same table?
* If two associations have the same name, how should the be referred to inside query clauses? (e.g. expressing conditions and select list of fields)
* How do we get different associations from the registry if they have the same name? (this is the problem you solved in this PR)
* What happens if in two different places a user decides to use the same registry alias for different associations?

**Comment by robertpustulka** [comment]
@lorenzo I'm not aiming to have two different associations with the same name in the same table - for me it would become problematic. `AssociationCollection` currently overwrites an association with the same name.

There are three aspects I want to address after some thoughts:

1) Association names are mutable, but that change isn't reflected in a collection nor the target table. Changing association name without manually reattaching it in a collection makes it unusable. `EagerLoader` [checks if these names match](https://github.com/cakephp/cakephp/blob/3.next/src/ORM/EagerLoader.php#L496) and throws an exception. Is this exception enough or should we disallow changing association name as it would probably break it.

2) Currently `className` option is ignored if we use the same association name more than once, note this example:
```php
//FooCategoriesTable
$table->hasMany('Children', [
    'className' => 'ChildFooCategories'
]);

//BarCategoriesTable
$table->hasMany('Children', [
    'className' => 'ChildBarCategories'
]);

$this->FooCategories->Children->getTarget(); //FooCategoriesTable
$this->BarCategories->Children->getTarget() //FooCategoriesTable !!!
```
I think that there should be at least an exception thrown if we try to bind an association with the same name but different `className` twice.

The constraint is a `TableLocator`/`TableRegistry` here. It does not allow having two different tables with the same alias in a registry, which is a reasonable thing. But `Association` [silently ignores](https://github.com/cakephp/cakephp/blob/3.next/src/ORM/Association.php#L399) the fact that we want to get a table with an alias that is already registered but have a different `className` ([contrary to `TableLocator`](https://github.com/cakephp/cakephp/blob/3.next/src/ORM/Locator/TableLocator.php#L171)). So there's another thing to improve.

3) The last thing is the original idea behind this PR: to have the same association name in different source tables and for different targets. Currently association name is identical to target table's alias. That's completely fine. But I was thinking if we could decouple association name from target table alias. But the more I think about it I realize that it would create more confusion. What alias should we use in conditions? What alias should be used in contain/matching? I think you're right @lorenzo that the current design is justified.

So that leaves points 1. and 2. to improve.

**Comment by lorenzo** [comment]
I agree that the association name should not be mutable once the association has been built.

And I also agree that trying to use the same association name more than once should create an exception, as long as we don't have a clever solution to using duplicates.

Would you like to take care of improving those 2 things?

**Comment by robertpustulka** [comment]
@lorenzo OK I can handle this. I'm wondering how to gracefully and BC-safe disable name setter.

File: src/ORM/Association.php [hunk 4]
```diff
@@ -101,6 +101,13 @@
     protected $_className;
 
     /**
+     * The registry alias of the target table object
+     *
+     * @var string
```
Comment (dereuromark) [inline src/ORM/Association.php]: this can also be |null

Comment (robertpustulka) [inline src/ORM/Association.php]: @dereuromark I'm dropping the idea as it leaves nowhere good. See comments above.

**Comment by lorenzo** [comment]
@robertpustulka maybe check if the name is not null in the setter and throw exception in that case?

**Comment by robertpustulka** [comment]
@lorenzo in that case we would allow for nullable names. Association name is set in a constructor and it shouldn't be changed afterwards. That means no setter at all.

The "soft" way to do this could be disabling a setter after the target table has been resolved. Before that developers could change association's name and reattach it in a collection with a proper key. After that changing association name would mean they would have to change both target table's alias (or target table instance) and a collection key...

Another way could be to check if a new association name matches the target table alias *after* it has been resolved.

In other words `AssociationCollection` key, `Association` name and target `Table` alias should be identical. Changing one of them would probably break things. And a question is where is the point we should prevent this from happenening.

**Comment by robertpustulka** [comment]
Issues has been addressed in #10154 and #10161

**Pull request closed without merging**